syntax = "proto2";
package tzrec.protos;

import "tzrec/protos/module.proto";
import "tzrec/protos/tower.proto";

enum Similarity {
  COSINE = 0;
  INNER_PRODUCT = 1;
}

message DSSM {
    required Tower user_tower = 1;
    required Tower item_tower = 2;
    // user and item tower output dimension
    required int32 output_dim = 3;
    // similarity method
    optional Similarity similarity = 4 [default=INNER_PRODUCT];
    // similarity scaling factor
    optional float temperature = 5 [default = 1.0];
    // use in batch items as negative items.
    optional bool in_batch_negative = 6 [default = false];
}

message DSSMV2 {
    required Tower user_tower = 1;
    required Tower item_tower = 2;
    // user and item tower output dimension
    required int32 output_dim = 3;
    // similarity method
    optional Similarity similarity = 4 [default=INNER_PRODUCT];
    // similarity scaling factor
    optional float temperature = 5 [default = 1.0];
    // use in batch items as negative items.
    optional bool in_batch_negative = 6 [default = false];
}

message TDM {
    required MultiWindowDIN multiwindow_din =1;
    required MLP final = 2;
}

message UserTowerPermFeat {
    required string tower_name = 1;
    repeated string features = 2;
    optional int32 nflods = 3 [default = 1];
}

message FusionDSSM {
    repeated Tower user_tower = 1;
    required Tower item_tower = 2;
    // user and item tower output dimension
    optional int32 output_dim = 3;
    optional Similarity similarity = 4 [default=INNER_PRODUCT];
    optional float temperature = 5 [default=0.1];
    optional bool use_senet = 6 [default=false];
    repeated UserTowerPermFeat user_tower_perm_feat = 7;
    // use in batch items as negative items.
    optional bool in_batch_negative = 8 [default = false];

}

message FusionDSSMV2 {
    repeated Tower user_tower = 1;
    required Tower item_tower = 2;
    // user and item tower output dimension
    optional int32 output_dim = 3;
    optional Similarity similarity = 4 [default=INNER_PRODUCT];
    optional float temperature = 5 [default=0.1];
    optional bool use_senet = 6 [default=false];
    repeated UserTowerPermFeat user_tower_perm_feat = 7;
    // use in batch items as negative items.
    optional bool in_batch_negative = 8 [default = false];
}
